sqlmesh/core/engine_adapter/duckdb.py

```diff
@@ -3,28 +3,41 @@
 import math
 import typing as t
 
-import pandas as pd
 from sqlglot import exp
 
+from sqlmesh.core.engine_adapter.base import SourceQuery
 from sqlmesh.core.engine_adapter.mixins import LogicalMergeMixin
 from sqlmesh.core.engine_adapter.shared import DataObject, DataObjectType
 
 if t.TYPE_CHECKING:
     from sqlmesh.core._typing import TableName
+    from sqlmesh.core.engine_adapter._typing import DF
 
 
 class DuckDBEngineAdapter(LogicalMergeMixin):
     DIALECT = "duckdb"
 
-    def _insert_append_pandas_df(
+    def _df_to_source_queries(
         self,
-        table_name: TableName,
-        df: pd.DataFrame,
-        columns_to_types: t.Optional[t.Dict[str, exp.DataType]] = None,
-        contains_json: bool = False,
-    ) -> None:
-        column_names = list(columns_to_types or [])
-        self.execute(exp.insert("SELECT * FROM df", table_name, columns=column_names))
+        df: DF,
+        columns_to_types: t.Optional[t.Dict[str, exp.DataType]],
+        batch_size: int,
+        target_table: TableName,
+    ) -> t.List[SourceQuery]:
+        assert columns_to_types
+        temp_table = self._get_temp_table(target_table)
+        casted_columns = [
+            exp.alias_(exp.cast(column, to=kind), column, copy=False)
+            for column, kind in columns_to_types.items()
+        ]
+        temp_table_sql = exp.select(*casted_columns).from_("df").sql(dialect=self.dialect)
```

Comment:
why do you need a temp table?

Reply:
If you later execute a query that has `select * from df` then it requires that `df` be defined within that scope. So a workaround for this could be to put `df` in globals but similar logic to other engines it should cheap to just make a temp table since dataframes can only be so big.